### ISSUE TITLE
setIndex should not set page to 0

### ISSUE THREAD

**Issue by maxiloc** [algolia/algoliasearch-helper-js#121]
I understand why it was there in the first place but it the snippet bellow will not load page 2. And I think it is not what we want.

```
helper.setCurrentPage(2);
helper.setIndex(indexName);
helper.search()
```

**Comment by vvo** [comment]
PR or it did not happen :dancer:

**Comment by bobylito** [comment]
I don't agree, while changing the index you can't know in advance how many pages there will be in the next search therefore resetting the page to 0 is a safe bet. If you are sure that there will be a page, let's say because you are using a slave that only changes the ordering of the results, then you are in a specific case. And then you can use : 

```
var page = helper.getCurrentPage();
helper.setIndex( indexName )
      .setCurrentPage( page )
      .search();
```

**Comment by vvo** [comment]
I agree with both arguments given their context but still @maxiloc was trapped by the behavior of setIndex when loading a state from a url, the order you do .setCurrentPage and .setIndex should not matter on the result.

If @maxiloc was trapped then I guess other people will be trapped too?

So:
- when changing the index using a UX element the expected behavior is always to be at page 0. Even if you were at page 5 and you are sure there will be a page 5 in the new index, it doesn't make sense to keep the page between the sort changes, no one would expect this.
- when there's a pageload, the .setCurrentPage and .setIndex order should not matter because it's disturbing

There's no API to handle both cases easily so I would just not set the page in setIndex because setIndex on a API level says: "set the index". not "set the index and the page".

Switching to page 0 when changing the index is a sweet shortcut but it can be tricky in some situations (page load).

We rather let the API consumer figure out that yes if he's implementing a sorting dropdown, he should reset to page 0 before calling search.

thoughts?

**Comment by bobylito** [comment]
Ok I understand better now. Makes sense when using this API. But, there is
a cleaner way to do that when loading from a URL. Using setState, first
retrieve the partial from the URL into an object, and then use setState to
fix the state as defined in the URL.

**Comment by redox** [comment]
I understand your point @bobylito but I think there are a lot of other things you can do that will generate a 0 results answer and that are not checked -> like if I add a wrong numeric filter and so on.

I would remove that `setCurrentPage` from the `setIndex` method to make sure we don't do anything magic.

**Comment by bobylito** [comment]
In the case of the URL I would still suggest using the setState, because it's less mutations and less functions calls.

But if you think that there is magic when we set the current page to 0 when we change the index, we have to have the same behavior in every methods that has the same behavior. Will fix that in the next version, but this will break the code of the users of the current version.

**Comment by bobylito** [comment]
Why don't we discuss about this issue around a cup of coffee Monday morning?

**Comment by vvo** [comment]
Of course we will :)

**Comment by Nainterceptor** [comment]
Hello,
I have an issue, probably linked to this one.

My fixed workflow call right my page :
* `helper.clearRefinements`
* `helper.setQueryParameter`
* `helper.addFacetRefinement`
* `helper.addDisjunctiveFacetRefinement`
* `helper.on('result', function(){})`
* __`helper.setCurrentPage`__
* `helper.search`

But in this order, it's not working : 
* __`helper.setCurrentPage`__
* `helper.clearRefinements`
* `helper.setQueryParameter`
* `helper.addFacetRefinement`
* `helper.addDisjunctiveFacetRefinement`
* `helper.on('result', function(){})`
* `helper.search`

Maybe it's important to document somewhere _when_ we should call setCurrentPage  ?

Have a nice day,
Gaël

**Comment by bobylito** [comment]
@Nainterceptor Any method that will change the number of pages will reset the current page to avoid inconsistencies. For example, if it was not the case we could stay on page 3 even though there were only 2 pages returned. So in fact, in your example `clearRefinements` `setQueryParameter` `addFacetRefinement` and `addDisjunctiveFacetRefinement` will set the current page to 0. You should call `setCurrentPage` when you want to change the current page.

**Comment by Nainterceptor** [comment]
@bobylito It make sense, I agree, but it's maybe important to write somewhere in the doc, to prevent bad usages like mine ?

**Comment by bobylito** [comment]
You are right! :) #274

**Comment by vvo** [comment]
Maybe we could be even more magical and solve this issue by not resetting to page 0 if the page was manually set somewhere before a call to .search.

So if at any moment the user specifically says he wants page 1 and calls for .search() then he should be one page 1, no matter the calling order.

Documenting it will maybe help people struggling with this issue get an answer at some point. But only after having struggled with it for a good amount of time.

If we can fix the issue for good we should do it?

**Comment by rayrutjes** [comment]
+1 for me on this one, this currently breaks pagination on all WordPress instant-search instances ;)

**Comment by vvo** [comment]
Some context given @rayrutjes comment: https://github.com/algolia/algoliasearch-wordpress/pull/305

I know this is not constructive to say "We need to fix this". We just experienced a really painful debugging session because of the setPage(0) being done, behind the scenes.

**Comment by olance** [comment]
+1, just spent ~1h debugging instantsearch pagination wondering why an out-of-the-box widget wouldn't change pages... only to realize I have some code that changes the helper state for some advanced use case and was resetting the page before `search()` was called.

Although I understand where it comes from, it's an unexpected side-effect and breaks the principle of least astonishment. (yeah, big words! :))

**Comment by Jerskouille** [comment]
I also got bitten by `setQueryParameter`'s `setPage(0)`.
I used an init widget only to change the `highlight{Pre,Post}Tag` s and couldn't understand why it was failing.

In my case, I guess I could just fix it by using `instantsearch.js`'s `searchParams` instead, but it definitely is surprising.

**Comment by vvo** [comment]
Since we are gonna work on a revamped helper version those feedbacks are important and shows that everytime we do a little bit of magic in low level libraries it's worth considering the impacts it has more precisely.

Thanks!

**Comment by rayrutjes** [comment]
We need to methods for this indeed ;)

**Comment by Jerskouille** [comment]
Spent half an hour again today on this one in Zendesk. /o\
I was used to using custom widgets to set my refinements and did a hack to set a query parameter in the searchFunction, both of which I had to surround with `cont page = getPage(); ...; setPage(page)`.

**Comment by Jerskouille** [comment]
Basically, I think the idea here is that it doesn't make sense to reset the page when done before the `search.start()` of `instantsearch.js`, but it does when done afterwards. Maybe worth having a before start / after start mode ?

**Comment by vvo** [comment]
I would say for such a low level library, when modifying a state object, we should never try to be smart on the page being reset to 0. We have now seen that many people were tricked by it.

We are gonna work on instantsearch-core in January and will tackle this issue inside it I think.

**Comment by bobylito** [comment]
Alright this is indeed very not cool what you have to implement @Jerskouille :/ Let's move forward.

**Comment by Jerskouille** [comment]
Getting bit in my hack inside the `searchFunction` is definitely OK, the surpising part is really inside an `init` function of a custom widget.

**Comment by tekstrand** [comment]
Wanted to check in to see where this is at and if anyone has a recommended workaround. I'm running into this while try to create an initial query object based off of URL parameters in my app.

**Comment by Haroenv** [comment]
That’s still the case. You need to `getPage` first and do `setPage` with that page after doing that mutation.

**Comment by drzraf** [comment]
+1. I spent hours figuring out what was happening.
I based my `searchFunction` on what is provided by https://www.algolia.com/doc/api-reference/widgets/instantsearch/js/#widget-param-searchfunction
But even though the comments state `// we re-apply the previous page`, it's so counter-intuitive and hard to understand!
1. It appears like the button (`view more`, `page 2`) does not work.
2. You need to disable caching to see actual (redundants) requests issued and see `page=0`
3. You need to understand why page isn't incremented (and find this ticket or read X times the above link)

Some clearer warning must be set somewhere about this `setPage()` usage within `searchFunction`.
